superdesk/core/storage.py

```
@@ -0,0 +1,261 @@
+
+    async def delete(self, file_id: Union[ObjectId, Any], resource: Optional[str] = None) -> None:
+        """
+        Delete a file from GridFS by its file ID.
+
+        :param file_id: The ID of the file to delete. This can be either a bson.ObjectId or any type
+                        that can be converted to a bson.ObjectId.
+        :param resource: The resource to use. Defaults to "upload" if not specified.
+
+        .. note:: Deletes of non-existent files are considered successful since the end result is the same.
+        """
+        media_id = ObjectId(file_id) if ObjectId.is_valid(file_id) else file_id
+        fs = await self.fs(resource)
+
+        try:
+            await fs.delete(media_id)
+        except gridfs.errors.NoFile:
+            logger.info(f"File with id: {file_id} was not found")
+
+    async def get_by_filename(self, filename: str, resource: Optional[str] = None) -> Optional[AsyncIOMotorGridOut]:
+        """
+        Retrieve a file from GridFS by its filename.
+
+        :param filename: The filename of the file to retrieve. The filename is expected to include
+                        an extension that will be removed to get the file ID.
+        :param resource: The resource to use. Defaults to "upload" if not specified.
+        :return: The file object retrieved from GridFS, or None if the file does not exist.
+        """
+        file_id, _ = os.path.splitext(filename)
+        return await self.get(file_id, resource)
+
+    async def fs(self, resource: Optional[str] = None) -> AsyncIOMotorGridFSBucket:
+        """
+        Get the GridFS bucket for the given resource.
+
+        :param resource: The resource to use. Defaults to "upload".
+        :return: The GridFS bucket.
+        """
+        resource = resource or "upload"
+        mongo = get_current_async_app().mongo
+        px = mongo.get_resource_config(resource).prefix
+
+        if px not in self._fs:
+            _, db = mongo.get_client_async(resource)
```

I was thinking here, if `resource` is not registered with the new system, should we fallback to the default mongo connection config (like we're now doing with the [MongoLock](https://github.com/superdesk/superdesk-core/blob/async/superdesk/lock.py#L46))?

We could easily do this by wrapping this call into a `try...except KeyError` block, and in the exception code block get the default config using `get_mong-client_config`. This would also mean you don't need the upload resource in the tests.

Thoughts?

---

An improvement might be to add a `get_default_client_async` function onto the [MongoResources](https://github.com/superdesk/superdesk-core/blob/async/superdesk/core/mongo.py#L172) class that will return a mongo client/db, and use that here (and in the superdesk.lock module)

---

I would rather raise an error in that case, this could be error prone when you have a typo in resource you might get different connection, most of the time we get the same but in those few times when it matters it might be doing something unexpected

---

I think I agree with the idea of raising an error instead and let the developer handle it.